Mathlib/Data/Pi/Projections.lean

```diff
@@ -0,0 +1,232 @@
+/-
+Copyright (c) 2024 Etienne Marion. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Etienne Marion
+-/
+import Mathlib.MeasureTheory.MeasurableSpace.Basic
+import Mathlib.Order.Interval.Finset.Nat
+import Mathlib.Topology.Constructions
+
+/-!
+# Projection of a function
+
+Given a dependent function `f : (i : ι) → α i` and `s : Set ι`, one might want to consider
+the restriction of `f` to the variables contained in `s`.
+This is defined in this file as `Function.proj s f`.
+Similarly, if we have `s t : Set ι`, `hst : s ⊆ t` and `f : (i : ↑t) → α ↑i`,
+one might want to restrict it to variables contained in `s`.
+This is defined in this file as `Function.proj₂ hst f`.
+We also define analoguous functions when `Finset`s are provided.
+
+Having these definitions avoids heavy type ascription when manipulating these functions and allow
+some rewriting when composing them (see `Function.proj₂_comp_proj` for instance).
+
+We also define analoguous functions for sets of the form `Set.Iic n` and `Finset.Iic n`, where
+`n : ℕ`, see `Function.projNat` for instance.
+
+## Main definitions
+
+* `Function.proj s f`: Restricts the function `f` to the variables contained in `s`.
+* `Function.projNat`: Restricts the function `f` to the variables indexed by integers
+  smaller than `n`.
+-/
+
+open MeasureTheory Function
+
+section Generality
+
+variable {ι : Type*}
+
+namespace Function
+
+variable {α : ι → Type*}
+
+/-- Given a dependent function, restrict it to a function of variables in `s`. -/
+@[simp]
+def proj (s : Set ι) (x : (i : ι) → α i) (i : s) : α i := x i
+
+/-- Given a dependent function of variables in `t`, restrict it to a function of variables in `s`
+when `s ⊆ t`. -/
+@[simp]
+def proj₂ {s t : Set ι} (hst : s ⊆ t) (x : (i : t) → α i) (i : s) : α i := x ⟨i.1, hst i.2⟩
+
+/-- Given a dependent function, restrict it to a function of variables in `s`, `Finset` version. -/
+@[simp]
+def fproj (s : Finset ι) (x : (i : ι) → α i) (i : s) : α i := x i
```

Comment:
Do you really need this one, and all the other finset guys, or is it enough to work with sets and the coercion from finsets to sets?

Reply:
I think I do. Let me explain why.

The main issue I have is with `proj₂`. If I have `s t : Finset i` and `hst : s \subset t`, then `proj₂ hst` does weird things. The problem is typically that `hst` is interpreted as an inclusion of multisets (which is the definition), and this causes trouble when doing some rewrite with `proj₂_comp_proj` for instance. So I would have to write `Finset.coe_subset.2 hst` each time, which does not seem ideal.

So I thought I would have to add `fproj₂`. Now consider the following mwe:
```lean
import Mathlib

open MeasureTheory

variable {ι : Type*} {α : ι → Type*} [∀ i, MeasurableSpace (α i)]

def proj (s : Set ι) (x : (i : ι) → α i) (i : s) : α i := x i

def fproj₂ {s t : Finset ι} (hst : s ⊆ t) (x : (i : t) → α i) (i : s) : α i := x ⟨i.1, hst i.2⟩

theorem fproj₂_comp_proj {s t : Finset ι} (hst : s ⊆ t) :
    (fproj₂ hst) ∘ (@proj ι α t) = proj s.toSet := rfl

theorem measurable_proj (s : Set ι) : Measurable (@proj ι α s) :=
  measurable_pi_lambda _ fun _ ↦ measurable_pi_apply _

theorem measurable_fproj₂ {s t : Finset ι} (hst : s ⊆ t) :
    Measurable (fproj₂ (α := α) hst) :=
  measurable_pi_lambda _ fun _ ↦ measurable_pi_apply _

example (s t : Finset ι) (hst : s ⊆ t) (μ : Measure (Π i, α i)) : μ.map (@proj ι α s) = sorry := by
  rewrite [← fproj₂_comp_proj hst, ← Measure.map_map (measurable_fproj₂ _) (measurable_proj _)]
  sorry
```

Here `rewrite` does not recognize `proj` when I write `measurable_proj _`. I know this tactic can't see through definitions, but in that case I find it very annoying having to specify the set. I don't really understand what happens here but it's annoying. So I thought I would also add `fproj`.
In my project I'm only dealing with finsets, and though `↑↑s : Type` is defeq to `↑s : Type`, the former seems to create some difficulties, typically asking for a `rfl` after a `rewrite`, which does not happen with the latter.

For all those reasons, I think this version better. I agree it is a bit unsatisfying to have that kind of duplicate, but on the other hand it feels much more comfortable. Maybe I'm doing something wrong, it's a bit difficult to convey the difficulties I met because those are on tiny examples from previous tests I deleted, but please don't hesitate if you have some suggestions!